#! /bin/sh /usr/share/dpatch/dpatch-run
## 06_add_replaygain_preferred_option.dpatch by dodamn <dodamn@gmail.com>
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: add 'track-preferred', 'album-preferred' values for replaygain options 
## DP: in configuration file

@DPATCH@

diff --git a/options.c b/options.c
index af80821..0797a93 100644
--- a/options.c
+++ b/options.c
@@ -535,7 +535,7 @@ static void toggle_repeat(unsigned int id)
 }
 
 const char * const replaygain_names[] = {
-	"disabled", "track", "album", NULL
+	"disabled", "track", "album", "track-preferred", "album-preferred", NULL
 };
 
 static void get_replaygain(unsigned int id, char *buf)
@@ -547,14 +547,14 @@ static void set_replaygain(unsigned int id, const char *buf)
 {
 	int tmp;
 
-	if (!parse_enum(buf, 0, 2, replaygain_names, &tmp))
+	if (!parse_enum(buf, 0, 4, replaygain_names, &tmp))
 		return;
 	player_set_rg(tmp);
 }
 
 static void toggle_replaygain(unsigned int id)
 {
-	player_set_rg((replaygain + 1) % 3);
+	player_set_rg((replaygain + 1) % 5);
 }
 
 static void get_replaygain_limit(unsigned int id, char *buf)
diff --git a/player.c b/player.c
index dcbeb5f..5d83cd9 100644
--- a/player.c
+++ b/player.c
@@ -233,7 +233,7 @@ static void update_rg_scale(void)
 	if (!player_info.ti || !replaygain)
 		return;
 
-	if (replaygain == RG_TRACK) {
+	if (replaygain == RG_TRACK || replaygain == RG_TRACK_PREFERRED) {
 		g = comments_get_val(player_info.ti->comments, "replaygain_track_gain");
 		p = comments_get_val(player_info.ti->comments, "replaygain_track_peak");
 	} else {
@@ -241,6 +241,16 @@ static void update_rg_scale(void)
 		p = comments_get_val(player_info.ti->comments, "replaygain_album_peak");
 	}
 
+    if (!g || !p) {
+        if (replaygain == RG_TRACK_PREFERRED) {
+    		g = comments_get_val(player_info.ti->comments, "replaygain_album_gain");
+	    	p = comments_get_val(player_info.ti->comments, "replaygain_album_peak");
+        } else if (replaygain == RG_ALBUM_PREFERRED) {
+    		g = comments_get_val(player_info.ti->comments, "replaygain_track_gain");
+	    	p = comments_get_val(player_info.ti->comments, "replaygain_track_peak");
+        }
+    }
+
 	if (!g || !p) {
 		d_print("gain or peak not available\n");
 		return;
diff --git a/player.h b/player.h
index f118551..acdaff0 100644
--- a/player.h
+++ b/player.h
@@ -44,7 +44,9 @@ enum player_status {
 enum replaygain {
 	RG_DISABLED,
 	RG_TRACK,
-	RG_ALBUM
+	RG_ALBUM,
+    RG_TRACK_PREFERRED,
+    RG_ALBUM_PREFERRED
 };
 
 struct player_callbacks {
